engine/consensus/matching/engine.go

```diff
@@ -343,13 +362,26 @@ func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionRece
 	// finalizer when blocks are added to the chain, and the IncorporatedBlockID
 	// will be the ID of the first block on its fork that contains a receipt
 	// committing to this result.
-	err = e.storeIncorporatedResult(receipt, &log)
+	added, err := e.storeIncorporatedResult(receipt)
 	if err != nil {
-		if engine.IsDuplicatedEntryError(err) {
-			return nil
-		}
 		return fmt.Errorf("failed to store incorporated result: %w", err)
 	}
+
+	if !added {
+		return nil
+	}
+
+	previousResultID := receipt.ExecutionResult.PreviousResultID
+	childReceipt, ok := e.pendingReceipts.ByPreviousResultID(previousResultID)
+	// if the receipt has a child, we will process it.
+	// and it's guaranteed won't go to the pending receipt again,
+	// because we've stored its parent receipt.
+	if ok {
+		e.SubmitLocal(childReceipt)
+	}
```

Comment:
I think adding this logic at his point adds complications to the structural improvements, which Yurii is working on: https://github.com/onflow/flow-go/pull/355 (essentially, this logic will be run in a threaded context, where submit is not available). 

I would suggest we follow a similar approach here as we did with HotStuff: 
* the caching logic for receipts with missing parents is in my opinion unrelated to the core business logic of the matching engine. 
* I think we should separate them. The matching engine only processes receipts whose parent result is attached to the Execution Tree.

Reply:
Sorry, I'm not clear about your suggestion? Are you suggesting to remove the pending receipts pool?

Reply:
I guess what you mean, is to follow the similar approach in [the compliance engine to continue processing](https://github.com/onflow/flow-go/blob/master/engine/consensus/compliance/engine.go#L525) the pending children proposals in the same thread?

Yeah, I was thinking about that too, and agree that's an efficient approach. Just that it might mess up the metrics. But I think I have just [figured a solution to that ](https://github.com/onflow/flow-go/pull/366#discussion_r573080129)